go-controller/pkg/node/vrfmanager/vrf_manager.go

```diff
@@ -0,0 +1,346 @@
+// with -vrf.
+func (vrfm *Controller) Repair(validVRFs sets.Set[string]) error {
+	vrfm.mu.Lock()
+	defer vrfm.mu.Unlock()
+
+	links, err := util.GetNetLinkOps().LinkList()
+	if err != nil {
+		return fmt.Errorf("failed to list links on the node, err: %v", err)
+	}
+
+	for _, link := range links {
+		name := link.Attrs().Name
+		// Skip if the link is not a vrf type or name is not suffixed with -vrf.
+		if link.Type() != "vrf" || !strings.HasSuffix(name, types.VRFDeviceSuffix) {
+			continue
+		}
+		if !validVRFs.Has(name) {
+			err = util.GetNetLinkOps().LinkDelete(link)
+			if err != nil {
+				klog.Errorf("VRF Manager: error deleting stale VRF device %s, err: %v", name, err)
+			}
+		}
+		delete(vrfm.vrfs, name)
+	}
+	return nil
+}
+
+// DeleteVRF deletes given VRF device from the node.
+func (vrfm *Controller) DeleteVRF(name string) (err error) {
+	vrfm.mu.Lock()
+	defer func() {
+		if err == nil {
+			delete(vrfm.vrfs, name)
+		}
+		vrfm.mu.Unlock()
+	}()
+	vrf, ok := vrfm.vrfs[name]
+	if !ok {
+		klog.V(5).Infof("VRF Manager: VRF %s not found in cache for deletion", name)
+		return nil
+	}
+	vrf.deleted = true
+	err = vrfm.deleteVRF(vrf)
+	if err != nil {
+		return fmt.Errorf("failed to delete VRF device %s, err: %w", vrf.name, err)
+	}
+	return nil
+}
+
+func (vrfm *Controller) deleteVRF(vrf vrf) error {
+	link, err := util.GetNetLinkOps().LinkByName(vrf.name)
+	if err != nil && util.GetNetLinkOps().IsLinkNotFoundError(err) {
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("failed to retrieve VRF device %s, err: %v", vrf.name, err)
+	}
+	return util.GetNetLinkOps().LinkDelete(link)
+}
+
+func getSlaveInterfaceNamesForVRF(vrfLink netlink.Link) (sets.Set[string], error) {
+	links, err := util.GetNetLinkOps().LinkList()
+	if err != nil {
+		return nil, fmt.Errorf("failed to list links on the node, err: %v", err)
+	}
+	enslavedInterfaces := make(sets.Set[string])
+	for _, link := range links {
+		if link.Attrs().MasterIndex == vrfLink.Attrs().Index {
+			enslavedInterfaces.Insert(link.Attrs().Name)
+		}
+	}
+	return enslavedInterfaces, nil
+}
+
+func enslaveInterfaceToVRF(vrfName, ifName string) error {
+	iface, err := util.GetNetLinkOps().LinkByName(ifName)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve interface %s, err: %v", ifName, err)
+	}
+	vrfLink, err := util.GetNetLinkOps().LinkByName(vrfName)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve VRF device %s, err: %v", vrfName, err)
+	}
+	err = util.GetNetLinkOps().LinkSetMaster(iface, vrfLink)
+	if err != nil {
+		return fmt.Errorf("failed to enslave interface %s to VRF %s: %v", ifName, vrfName, err)
+	}
+	return nil
+}
+
+func removeInterfaceFromVRF(vrfName, ifName string) error {
+	iface, err := util.GetNetLinkOps().LinkByName(ifName)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve interface %s, err: %v", ifName, err)
+	}
+	err = util.GetNetLinkOps().LinkSetMaster(iface, nil)
```

Comment:
will this actually work? I see in LinkSetMaster:

```
// LinkSetMaster sets the master of the link device.
// Equivalent to: `ip link set $link master $master`
func (h *Handle) LinkSetMaster(link Link, master Link) error {
	index := 0
	if master != nil {
		masterBase := master.Attrs()
		h.ensureIndex(masterBase)
		index = masterBase.Index
	}
	if index <= 0 {
		return fmt.Errorf("Device does not exist")
	}
	return h.LinkSetMasterByIndex(link, index)
}
```
Should hit "Device doesnt exist" error

Reply:
Think you want `LinkSetNoMaster`

Reply:
this was always invoked when VRF link is present, but it is a good find, awesome.
changed to use `LinkSetNoMaster`.